packages/go_router/test/builder_test.dart

```diff
@@ -369,6 +370,45 @@ void main() {
           (shellNavigatorKey.currentWidget as Navigator?)?.restorationScopeId,
           'scope1');
     });
+
+    testWidgets('RouteBuilderException gets catched',
+        (WidgetTester tester) async {
+      final RouteConfiguration config = RouteConfiguration(
+        routes: <RouteBase>[
+          GoRoute(
+            path: '/',
+            builder: (BuildContext context, GoRouterState state) {
+              throw RouteBuilderException('Something went wrong!');
+            },
+          ),
+        ],
+        redirectLimit: 10,
+        topRedirect: (BuildContext context, GoRouterState state) => null,
+        navigatorKey: GlobalKey<NavigatorState>(),
+      );
+
+      final RouteMatchList matches = RouteMatchList(
+          matches: <RouteMatch>[
+            RouteMatch(
+              route: config.routes.first as GoRoute,
+              matchedLocation: '/',
+              extra: null,
+              error: null,
+              pageKey: const ValueKey<String>('/'),
+            ),
+          ],
+          uri: Uri.parse('/'),
+          pathParameters: const <String, String>{});
+
+      await tester.pumpWidget(
+        _BuilderTestWidget(
+          routeConfiguration: config,
+          matches: matches,
+        ),
+      );
+      expect(tester.takeException(), isInstanceOf<RouteBuilderException>());
```

Comment:
You won't be able to take exception because it was caught internally by the RouteBuilder

Reply:
That's what I thought too, but if I remove this line the test fails because the test framework catches the error, so strange.

Reply:
Ah I see, the builder error is only caught during the page building part. so the builder method in the goRoute is not actually called during this part.

The error is throw when flutter draws frame. and it will be another async process that go_router can't catch. I don't think the thing you are trying to achieve here is possible...

What are your use cases?

Reply:
I would like the error builder to build when an RouteBuilderException is thrown in a route builder.

Reply:
Using this PR it works fine, the when a RouteBuilderException is thrown GoRouter calls the errorBuilder, the problem is finding a way to test it.

Reply:
for page builder it would work. But for builder it won't.

Based on the code, the RouteBuilderError should only be handling error in MatchRouteList.

In your use case, can you do a context.go if things went wrong in the builder?

Reply:
I was using the older version of GoRouter and this worked fine. I'm throwing custom exceptions depending on specific events that can occur during the build of the route and i'm using the error builder to show different error screens depending on the error thrown. 

Do you see a way of making this work for builder too ?

Reply:
I don't think it was meant to be used this way. Since the entire build pipeline is divided between go_router and flutter. and some of the refactoring probably move some part of the build from go_router to flutter side.

I can't see a way to fix this without corner cases. 

> I'm throwing custom exceptions depending on specific events that can occur during the build of the route

I think the best bet is to do a context.go when the condition is met. probably in a postframecallback. or just build a different widget base on that.